cli/main.c

```diff
@@ -90,6 +90,17 @@ int switchtec_handler(const char *optarg, void *value_addr,
 	return 0;
 }
 
+int mfg_handler(const char *optarg, void *value_addr,
+		const struct argconfig_options *opt)
+{
+	if(strstr(optarg, "tty")) {
```

Comment:
No, this is junk, the point was you weren't supposed to need to do this because as soon as you do a bad command it returns a sensible error.... Why are you adding this?

Reply:
The previous fix works for MAIN boot phase. 

However, in BL1/2, program fails at 'switchtec_open()', when it tries to read device ID by reading GAS register. BL1/2 does not support UART access of GAS so the function fails with 'Error communicating with device. Please check your setup.'

By adding this check, we block mfg access from CLI level, before switchtec_open().

From an API perspective, in BL1/2, since switchtec_open() of UART will fail, user will not have access to other API functions.

Reply:
Well then fix the actual problem, don't add hacks like this...

Something along these lines.... But I now suspect that the switchtec_mfg_cmd() is wrong and should be dropped in favor of something like this:

```diff
diff --git a/lib/platform/gasops.c b/lib/platform/gasops.c
index 629412face20..9c8e7fcf4710 100644
--- a/lib/platform/gasops.c
+++ b/lib/platform/gasops.c
@@ -104,7 +104,19 @@ int gasop_cmd(struct switchtec_dev *dev, uint32_t cmd,
 
 int gasop_get_device_id(struct switchtec_dev *dev)
 {
-       return gas_reg_read32(dev, sys_info.device_id);
+       int ret;
+
+       ret = gas_reg_read32(dev, sys_info.device_id);
+
+       if (ret == -1) {
+               if (dev->ops->flags & SWITCHTEC_OPS_FLAG_NO_BL)
+                       errno = ERR_IFACE_NOT_SUPPORTED_IN_BL;
+               else
+                       errno = EIO;
+       }
+
+       return ret;
 }
 
 int gasop_get_fw_version(struct switchtec_dev *dev, char *buf,
diff --git a/lib/switchtec.c b/lib/switchtec.c
index 395c4aa27924..2bb86c05a884 100755
--- a/lib/switchtec.c
+++ b/lib/switchtec.c
@@ -130,6 +130,8 @@ static int set_gen_variant(struct switchtec_dev * dev)
        dev->var = SWITCHTEC_VAR_UNKNOWN;
 
        dev->device_id = dev->ops->get_device_id(dev);
+       if (dev->device_id == -1)
+               return -1;
 
        while (id->device_id) {
                if (id->device_id == dev->device_id) {
```

Reply:
The previous switchtec_mfg_cmd() fix works for MAIN boot phase where the device ID can be read through UART, so it is a valid fix. 

In BL1/2, switchtec_open fails at uart_gas_read() where it does "raise(SIGBUS);", so the above patch will not work. 

Removing the "raise(SIGBUS)" will have a wider impact on other functions... So there is no quick solution. I'll look for another solution.

Reply:
Add a check_gas_access() op (or something like that) that's called in switchtec_open() and is skipped if not set. The uart platform would provide an implementation that tries a gas read and fails with an appropriate errno if it doesn't work.

This will require a bit of cleanup in the uart platform....

Reply:
Also at this layer, we won't know if a UART gas read fail is because actual communication issue, or because we are in BL1/2 (boot phase is determined AFTER gas read), so we cannot give user a meaningful error code.